### PR TITLE
Platform/13199/fhir function dynamic message routing

### PR DESCRIPTION
**Pull request by david-navapbc**
This PR ...

Refactors how queue message routing happens such that the behavior (1) no longer hard coded (2) baked into the queue message itself. ~~The trigger for changing default routing behavior is expected to be the something in the content of the message itself and thus it makes sense for the routing logic to be colocated with the content of the message itself.~~ 

## Changes
~~QueueMessage has been updated with a "when" clause (switch/case for Java-heads) that sets default routing behavior for a given queue message type. Should a short circuit of this behavior be required the germane "when" block for that message type may be updated to route the message to a different queue topic. the queueAccess object is provided to the a new function, QueueMessage.send() to keep separation of concern in tact.  The QueueMessage object handles routing itself with routing plumbing provided by FHIRFunctions.~~

`QueueMessage`, `FhirFunctions`, and associated tests have been updated. 

`QueueMessage` was changed to attach the correct queue name to the QueueMessage obj typed for that kind of QueueMessage. Functionally speaking, the queue message itself knows where it's supposed to go. 

In `FhirFunctions` the calls to `doRoute(), doConvert(), etc` have been consolidated into a single common method. The test updates were made to accommodate this change. 

### Testing
- [x ] Tested locally?
- [x ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?

### CODE REVIEW

**Review by arnejduranovic — CHANGES_REQUESTED** [review]
This PR appears to me to be very similar to the original code. We are essentially moving extra configuration from one file to another. Additionally, the original implementation has the translate queue names defined in the package space in FHIREngine, which is poor practice. This PR is on the right track, and we can go even further by exploring a solution that keeps the queue name tied to the class, and make it impossible to configure a new QueueMessage class without defining the queue name. See the proof of concept patch below that illustrates this idea:

[OOP_QMessage_Cleanup_POC.patch](https://github.com/user-attachments/files/16339597/OOP_QMessage_Cleanup_POC.patch)

**Comment by david-navapbc** [comment]
@arnejduranovic I looked at the patch and I could use a 10m huddle because I think I'm missing something. If I read this right the queue name becomes a hard-coded value tied to the message type which is what I thought we were getting away from. I tied the queue name to a "when" block so we could add logic as needed to short circuit default behavior based on message content. If it's baked into the data class for a given message type then isn't it still hard coded?

**Comment by david-navapbc** [comment]
copied from slack for observability 

> looking at the PR comment. I think I get what you're after now and yes I misunderstood the intent.
> I thought the ask was to make is so a - for example - a "route" queue message could be put on the "translation" queue.
> What I'm understanding now is that no we don't want that. What we want is for the "route" processing step to be able to produce different kinds of queue messages vs having the kind of queue message that results from that processing always being one thing.
> 
> NEW
> 
> [9:25]
> I may still need a huddle but assuming I've got :this-forever:  right let me go back and look at the code for a bit and think on it. I can then come to you with thoughtful intelligent questions and save us both some time.
> 
> 
> Arnej Duranović
>   9:26 AM
> What we want is for the "route" processing step to be able to produce different kinds of queue messages vs having the kind of queue message that results from that processing always being one thing. yep, this is correct
> 
> 
> davidholiday
>   [9:26 AM](https://nava.slack.com/archives/D06S7U58U1G/p1721748379364869)
> ok cool. apologies for the misunderstanding and I get the ask now
> :raised_hands:
> 1
> 
> 9:26
> lemmie stare at the code and see what I can come up with

**Review by brick-green — APPROVED** [review]
nice, now a convert queue message can only go to the convert step. These changes meet the requirements as I understand them. All tests pass. Looks good :)

File: prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionIntegrationTests.kt [hunk 1]
```diff
@@ -786,7 +786,8 @@ class FhirFunctionIntegrationTests {
 
             databaseAccess = ReportStreamTestDatabaseContainer.testDatabaseAccess
         )
-        fhirFunc.doRoute(queueMessage, 1, fhirEngine, actionHistory)
+        // fhirFunc.doRoute(queueMessage, 1, fhirEngine, actionHistory)
```
Comment (arnejduranovic) [inline prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionIntegrationTests.kt:789]: clean up comments?

File: prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionTests.kt [hunk 2]
```diff
@@ -244,12 +244,13 @@ class FhirFunctionTests {
             emptyMap(),
             emptyList()
         )
-        val message = FhirRouteQueueMessage(
+        val message = FhirTranslateQueueMessage(
```
Comment (arnejduranovic) [inline prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionTests.kt:247]: What's the story here?

File: prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt [hunk 3]
```diff
@@ -13,9 +13,17 @@ import gov.cdc.prime.router.ReportId
 import gov.cdc.prime.router.Topic
 import gov.cdc.prime.router.azure.BlobAccess
 import gov.cdc.prime.router.azure.Event
+import gov.cdc.prime.router.azure.QueueAccess
 import java.util.Base64
 import java.util.UUID
 
+const val elrConvertQueueName = "elr-fhir-convert"
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt:20]: our code is open source - we're exposing all of it to the world. Also I think for testing if nothing else it's useful to have these at the top level vs baked into a particular message class. 

Happy to make the change though if you feel strongly.

File: prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt [hunk 4]
```diff
@@ -96,6 +112,7 @@ data class FhirConvertQueueMessage(
     override val blobSubFolderName: String,
     override val topic: Topic,
     val schemaName: String = "",
+    override val messageQueueName: String = elrConvertQueueName,
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt:115]: are they not functionally equivalent? what's the difference that makes a difference in having the override messageQueueName in the argument list vs as part of the body? 

Happy to do whatever - just curious as to what the issue is

File: prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionIntegrationTests.kt (hunk 1 above)
Comment (david-navapbc) [reply prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionIntegrationTests.kt:789]: ack

File: prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionTests.kt (hunk 2 above)
Comment (david-navapbc) [reply prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionTests.kt:247]: the test logic wasn't testing what it was supposed to be testing

File: prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (arnejduranovic) [reply prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt:20]: We currently have a requirement (implicit) that specific queue messages should go to specifically named queues in Azure. So, we should "hardcode" these in the concrete queue message classes and make it so the caller cannot change the QueueName. I don't see testing or open source considerations outweighing this requirement in this case.

File: prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (arnejduranovic) [reply prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt:115]: The issue is the caller can pass in a different messageQueueName with the way you have it now and we do not want to allow that ability. When a queue message is designed, we know at design time what the name of the azure queue should be so we should prevent it from being changed by the caller -- which is what the snippet I posted does.

File: prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt]: ok done

File: prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/fhirengine/engine/QueueMessage.kt]: done

**Review by arnejduranovic — CHANGES_REQUESTED** [review]

**Comment by arnejduranovic** [comment]
replaced by https://github.com/CDCgov/prime-reportstream/pull/15613. Closing.

**Pull request closed without merging**